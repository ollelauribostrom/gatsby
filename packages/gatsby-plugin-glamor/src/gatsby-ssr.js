import React from "react"
import { renderToString } from "react-dom/server"
import { renderStaticOptimized } from "glamor/server"

export const replaceRenderer = ({
  bodyComponent,
  replaceBodyHTMLString,
  setHeadComponents,
}) => {
  let { html, css, ids } = renderStaticOptimized(() =>
    renderToString(bodyComponent)
  )

  replaceBodyHTMLString(html)

  setHeadComponents([
    <style
      id="glamor-styles"
      key="glamor-styles"
      dangerouslySetInnerHTML={{ __html: css }}
    />,
    <script
      id="glamor-ids"
      key="glamor-ids"
      dangerouslySetInnerHTML={{
        __html: `
        // <![CDATA[
        window._glamor = ${JSON.stringify(ids)}
        // ]]>
        `,
      }}
    />,
  ])
}
