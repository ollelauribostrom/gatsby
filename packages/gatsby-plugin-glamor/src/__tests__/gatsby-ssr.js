jest.mock(`react-dom/server`)
jest.mock(`glamor/server`)

import React from "react"
import ReactDOM from "react-dom/server"
import * as glamor from "glamor/server"
import { replaceRenderer } from "../gatsby-ssr"

const setup = () => {
  const bodyComponent = <div />
  const replaceBodyHTMLString = jest.fn()
  const setHeadComponents = jest.fn()

  glamor.renderStaticOptimized = jest.fn(fn => fn())
  ReactDOM.renderToString = jest.fn(() => {
    return {
      html: `mock-html`,
      css: `mock-css`,
      ids: `mock-ids`,
    }
  })

  return { bodyComponent, replaceBodyHTMLString, setHeadComponents }
}

describe(`gatsby-plugin-glamor`, () => {
  describe(`replaceRenderer`, () => {
    it(`renders the correct html`, () => {
      const args = setup()

      replaceRenderer(args)

      expect(ReactDOM.renderToString).toHaveBeenCalledTimes(1)
      expect(ReactDOM.renderToString).toHaveBeenCalledWith(args.bodyComponent)
      expect(args.replaceBodyHTMLString).toHaveBeenCalledWith(`mock-html`)
    })

    it(`sets the correct head components`, () => {
      const args = setup()

      replaceRenderer(args)

      expect(args.setHeadComponents).toHaveBeenCalledTimes(1)
      expect(args.setHeadComponents.mock.calls).toMatchSnapshot()
    })
  })
})
