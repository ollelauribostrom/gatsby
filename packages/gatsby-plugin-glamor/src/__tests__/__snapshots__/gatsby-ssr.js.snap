// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-glamor replaceRenderer sets the correct head components 1`] = `
Array [
  Array [
    Array [
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "mock-css",
          }
        }
        id="glamor-styles"
      />,
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "
        // <![CDATA[
        window._glamor = \\"mock-ids\\"
        // ]]>
        ",
          }
        }
        id="glamor-ids"
      />,
    ],
  ],
]
`;
